contracts/core/MatchExecutor.sol

```diff
@@ -0,0 +1,369 @@
+// SPDX-License-Identifier: MIT
+pragma solidity 0.8.14;
+
+import {Ownable} from '@openzeppelin/contracts/access/Ownable.sol';
+import {Pausable} from '@openzeppelin/contracts/security/Pausable.sol';
+import {IERC165} from '@openzeppelin/contracts/interfaces/IERC165.sol';
+import {IERC20} from '@openzeppelin/contracts/token/ERC20/IERC20.sol';
+import {IERC721} from '@openzeppelin/contracts/token/ERC721/IERC721.sol';
+import {IERC721Receiver} from '@openzeppelin/contracts/token/ERC721/IERC721Receiver.sol';
+
+import {IFlashLoanRecipient} from '../interfaces/IFlashLoanRecipient.sol';
+import {IBalancerVault} from '../interfaces/IBalancerVault.sol';
+import {MatchExecutorTypes} from '../libs/MatchExecutorTypes.sol';
+import {OrderTypes} from '../libs/OrderTypes.sol';
+import {IInfinityExchange} from '../interfaces/IInfinityExchange.sol';
+
+/**
+@title MatchExecutor
+@author Joe
+@notice The contract that is called to execute order matches
+*/
+contract MatchExecutor is IFlashLoanRecipient, IERC721Receiver, Ownable, Pausable {
+  /*//////////////////////////////////////////////////////////////
+                                ADDRESSES
+      //////////////////////////////////////////////////////////////*/
+
+  /// @notice The address of the EOA that acts as an intermediary in the brokerage process
+  address public intermediary;
+
+  IInfinityExchange public immutable exchange;
+
+  IBalancerVault public immutable vault;
+
+  /*//////////////////////////////////////////////////////////////
+                              EXCHANGE STATES
+      //////////////////////////////////////////////////////////////*/
+
+  /// @notice Mapping to keep track of which exchanges are enabled
+  mapping(address => bool) public payableContracts;
+
+  /*//////////////////////////////////////////////////////////////
+                                EVENTS
+      //////////////////////////////////////////////////////////////*/
+  event IntermediaryUpdated(address indexed intermediary);
+  event PayableContractUpdated(address indexed exchange, bool indexed isPayable);
+
+  /*//////////////////////////////////////////////////////////////
+                               CONSTRUCTOR
+    //////////////////////////////////////////////////////////////*/
+  constructor(
+    address _intermediary,
+    IBalancerVault _vault,
+    IInfinityExchange _exchange
+  ) {
+    _updateIntermediary(_intermediary);
+    vault = _vault;
+    exchange = _exchange;
+  }
+
+  function onERC721Received(
+    address,
+    address,
+    uint256,
+    bytes calldata
+  ) external pure returns (bytes4) {
+    return this.onERC721Received.selector;
+  }
+
+  /**
+   * @notice Set the specified contract to allow payable calls
+   * @param _payableContract The address to allow or disallow
+   * @param _isPayable The state to update the address to
+   */
+  function updatePayableContract(address _payableContract, bool _isPayable) external onlyOwner {
```

Comment:
remove this and add exchangedAdded and exchangeRemoved functions when you replace mapping with enu set